lsp-pwsh.el

```diff
@@ -0,0 +1,307 @@
+
+(lsp-register-client
+ (make-lsp-client
+  :new-connection (lsp-stdio-connection #'lsp-pwsh--command)
+  :major-modes lsp-pwsh--major-modes
+  :server-id 'pwsh-ls
+  :priority 1
+  :initialization-options #'lsp-pwsh--extra-init-params
+  :notification-handlers (lsp-ht ("powerShell/executionStatusChanged" 'ignore))
+  :initialized-fn (lambda (w)
+                    (with-lsp-workspace w
+                      (lsp--set-configuration
+                       (lsp-configuration-section "powershell"))))
+  ))
+
+(add-to-list 'lsp-language-id-configuration '(powershell-mode . "powershell"))
+
+;; Compatibility
+(if (fboundp 'company-lsp)
+    (advice-add 'company-tng--supress-post-completion
+                :after-while
+                (lambda (&rest _)
+                  (not (memq major-mode lsp-pwsh--major-modes)))
+                '((name . --force-post-completion))))
+
+(defun lsp-pwsh--filter-cr (str)
+  "Filter CR entities from STR."
+  (when (and (eq system-type 'windows-nt) str)
+    (replace-regexp-in-string "\r" "" str)))
+
+(advice-add 'lsp-ui-doc--extract :filter-return #'lsp-pwsh--filter-cr)
+(advice-add 'lsp-ui-sideline--format-info :filter-return #'lsp-pwsh--filter-cr)
+
+;;; Utils
+
+(defconst lsp-pwsh-unzip-script "%s -noprofile -noninteractive -nologo -ex bypass -command Expand-Archive -Path '%s' -DestinationPath '%s'"
+  "Powershell script to unzip vscode extension package file.")
+
+(defconst lsp-pwsh-editor-svcs-dl-script "%s -noprofile -noninteractive -nologo -ex bypass -command Invoke-WebRequest -UseBasicParsing -uri '%s' -outfile '%s'"
+  "Command executed via `shell-command' to download the latest PowerShellEditorServices release.")
+
+(defcustom lsp-pwsh-github-asset-url
+  "https://github.com/%s/%s/releases/latest/download/%s"
+  "GitHub latest asset template url."
+  :group 'lsp-pwsh
+  :type 'string)
+
+(defun lsp-pwsh--get-extension (url dest)
+  "Get extension from URL and extract to DEST."
+  (let ((temp-file (make-temp-file "ext" nil ".zip")))
+    ;; since we know it's installed, use powershell to download the file
+    ;; (and avoid url.el bugginess or additional libraries)
+    (shell-command (format lsp-pwsh-editor-svcs-dl-script lsp-pwsh-exe url temp-file))
+    (message "lsp-pwsh: Downloading done!")
+    (if (file-exists-p dest) (delete-directory dest 'recursive))
+    (shell-command (format lsp-pwsh-unzip-script lsp-pwsh-exe temp-file dest))
+    (message "lsp-pwsh: Finish unzip!")))
+
+(defun lsp-pwsh-setup (&optional forced)
+  "Downloading PowerShellEditorServices to `lsp-pwsh-dir'.
+FORCED if specified with prefix argument."
+  (interactive "P")
+  (let ((parent-dir (file-name-directory lsp-pwsh-dir)))
+    (unless (and (not forced) (file-exists-p parent-dir))
+      (lsp-pwsh--get-extension
+       (format lsp-pwsh-github-asset-url "PowerShell" "PowerShellEditorServices" "PowerShellEditorServices.zip")
+       parent-dir))))
+
+;; Download extension
+(lsp-pwsh-setup)
```

Comment:
This should be moved from top-level to probably to `lsp-pwsh--command`. FYI At some point, we will have a field in the `lsp--client` for download commands and lsp-mode will automatically manage that process.